cmd/kubeadm/app/util/staticpod/utils_test.go

```diff
@@ -37,43 +42,162 @@ func TestComponentResources(t *testing.T) {
 
 func TestComponentProbe(t *testing.T) {
 	var tests = []struct {
-		port   int
-		path   string
-		scheme v1.URIScheme
+		name      string
+		cfg       *kubeadmapi.MasterConfiguration
+		component string
+		port      int
+		path      string
+		scheme    v1.URIScheme
 	}{
 		{
-			port:   1,
-			path:   "foo",
-			scheme: v1.URISchemeHTTP,
+			name: "default apiserver advertise address with http",
+			cfg: &kubeadmapi.MasterConfiguration{
+				API: kubeadmapi.API{
+					AdvertiseAddress: "",
+				},
+			},
+			component: kubeadmconstants.KubeAPIServer,
+			port:      1,
+			path:      "foo",
+			scheme:    v1.URISchemeHTTP,
+		},
+		{
+			name: "default apiserver advertise address with https",
+			cfg: &kubeadmapi.MasterConfiguration{
+				API: kubeadmapi.API{
+					AdvertiseAddress: "",
+				},
+			},
+			component: kubeadmconstants.KubeAPIServer,
+			port:      2,
+			path:      "bar",
+			scheme:    v1.URISchemeHTTPS,
+		},
+		{
+			name: "valid ipv4 apiserver advertise address with http",
+			cfg: &kubeadmapi.MasterConfiguration{
+				API: kubeadmapi.API{
+					AdvertiseAddress: "1.2.3.4",
+				},
+			},
+			component: kubeadmconstants.KubeAPIServer,
+			port:      1,
+			path:      "foo",
+			scheme:    v1.URISchemeHTTP,
 		},
 		{
-			port:   2,
-			path:   "bar",
-			scheme: v1.URISchemeHTTPS,
+			name: "valid IPv4 scheduler probe",
+			cfg: &kubeadmapi.MasterConfiguration{
+				SchedulerExtraArgs: map[string]string{"address": "1.2.3.4"},
+			},
+			component: kubeadmconstants.KubeScheduler,
+			port:      1,
+			path:      "foo",
+			scheme:    v1.URISchemeHTTP,
+		},
+		{
+			name: "valid etcd probe using listen-client-urls IPv4 addresses",
+			cfg: &kubeadmapi.MasterConfiguration{
+				Etcd: kubeadmapi.Etcd{
+					ExtraArgs: map[string]string{
+						"listen-client-urls": "http://1.2.3.4:2379,http://4.3.2.1:2379"},
+				},
+			},
+			component: kubeadmconstants.Etcd,
+			port:      1,
+			path:      "foo",
+			scheme:    v1.URISchemeHTTP,
+		},
+		{
+			name: "valid IPv4 etcd probe using hostname for listen-client-urls",
+			cfg: &kubeadmapi.MasterConfiguration{
+				Etcd: kubeadmapi.Etcd{
+					ExtraArgs: map[string]string{
+						"listen-client-urls": "http://localhost:2379"},
+				},
+			},
+			component: kubeadmconstants.Etcd,
+			port:      1,
+			path:      "foo",
+			scheme:    v1.URISchemeHTTP,
 		},
 	}
 	for _, rt := range tests {
-		actual := ComponentProbe(rt.port, rt.path, rt.scheme)
+		actual := ComponentProbe(rt.cfg, rt.component, rt.port, rt.path, rt.scheme)
+		switch {
+		case rt.component == kubeadmconstants.KubeAPIServer:
```

Comment:
Instead of adding all this complicated logic (which is not very scalable if we add more test scenarios), how about adding a new `expected` field that we do an equality on?

Reply:
Good call. I'll update and resubmit the PR.